### ISSUE TITLE
Move autograph into a separate package and into a separate repo

### ISSUE THREAD

**Issue by KOLANICH** [tensorflow/tensorflow#23601]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: any
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: n/a
- **TensorFlow installed from (source or binary)**: any
- **TensorFlow version (use command below)**: any
- **Python version**: any
- **Bazel version (if compiling from source)**: any
- **GCC/Compiler version (if compiling from source)**: any
- **CUDA/cuDNN version**: any
- **GPU model and memory**: any
- **Exact command to reproduce**: n/a

### Describe the problem

I guess autograph should be moved into a separate package and repo. Because I guess it may be possible to add support of other computation graph frameworks there.

**Comment by mdanatg** [comment]
Indeed, and we plan to do that very soon. Due to time constraints, we've initially moved it inside TF core, and from there we'll extract the reusable parts into a standalone repo. We hope to have that set up by the end of this year.

**Comment by chunduriv** [comment]
@KOLANICH,

Sorry for the late response. The autograph moved to a [new repo](https://github.com/tensorflow/autograph) and it can be accessed as [tf.autograph](https://www.tensorflow.org/api_docs/python/tf/autograph). Thank you.

**Comment by KOLANICH** [comment]
Thank you, and everyone involved into it. :tada:

**Comment by KOLANICH** [comment]
>The autograph moved to a new repo and it can be accessed as tf.autograph. Thank you.

It hasn't been moved to a new repo:

>This repository contains tests and example code

**Comment by mdanatg** [comment]
Yes, this is still an open issue (I think @chunduriv was referring to autograph being available in its own namespace). Mainly because we need someone to set up and maintain a separate repo - otherwise, significant parts of the source code (tensorflow/python/autograph/pyct) are already independent of TF and could be mover rather mechanically.

**Comment by rmshaffer** [comment]
Commenting here for visibility. We have made a static copy of the `autograph` codebase for use in a framework which builds on `autograph` to convert Python functions into OpenQASM programs for quantum computations. We did not want to take a dependency on TensorFlow, so the quickest path forward was copying the codebase and making minor modifications to decouple it from TensorFlow. We would be very interested in consuming `autograph` as a standalone package, if it ever becomes available, and would be happy to help validate the work with our use case.

Here is our copy of the `autograph` codebase:
https://github.com/amazon-braket/amazon-braket-sdk-python-staging/tree/feature/autoqasm/src/braket/experimental/autoqasm/autograph

More details on how we are using it:
https://github.com/amazon-braket/amazon-braket-sdk-python-staging/blob/feature/autoqasm/src/braket/experimental/autoqasm/README.md

**Comment by josh146** [comment]
Hey all, having autograph available as a separate repo would also be super valuable/important from our end. We're in a similar position to @rmshaffer -- we're depending on Autograph in [Catalyst](https://github.com/pennylaneai/catalyst), but don't want to have to depend on the full TensorFlow installation for users of our package.

Would it be possible to split autograph out -- without the TensorFlow dependency -- into its own PyPI package? We would be very happy to help out here :)

**Comment by mdanatg** [comment]
It's certainly possible to split it out - autograph was designed with that in mind. Besides some dependencies in the high level API that accumulated as technical debt and should be easy to remove, only the operators are TF-specific, so as long as you have replacement for those (which I suspect you want anyway), a split should definitely be doable.

I'm not sure if there are any concrete plans to do such a split, so I'd recommend taking the lead and doing it.

**Comment by dime10** [comment]
Hello TensorFlow maintainers, I'm currently tackling the separation of TensorFlow and AutoGraph as a contribution to TF.

I have a question regarding the preferred code organization, for which I could see the following options:
- keep all code within the tensorflow repo, while separating out build & packaging processes etc.
  - Following this model, either of the below could still be taken as a follow up.
- keep all code within the tensorflow repo (as above), but also make a mirror available in a separate repository
  - I believe this was the model followed by [mlir-hlo](https://github.com/tensorflow/mlir-hlo), although that code has now moved from tensorflow to xla.
- _move_ everything to a different repository, for instance the existing [autograph repository](https://github.com/tensorflow/autograph)
  - I believe this is also the model followed by the [tensorflow estimator](https://github.com/tensorflow/estimator) package, which is also available as a separate [pip package](https://pypi.org/project/tensorflow-estimator/).

**Comment by mdanatg** [comment]
I think the last option makes most sense, at least in the long run, although a clean move would not be easy to achieve. It would probably be most practical to start with a fork, and continue improvements there, while in parallel configuring TF for an external dependency.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by dime10** [comment]
Hi @tilakrayal, thank you for checking in. I don't think the latest TensorFlow release is resolving this issue, which is about separately packaging and releasing one of the sub-components of TensorFlow.

**Comment by Venkat6871** [comment]
Hi @KOLANICH ,
Thank you so much for filing this feature request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!

**Comment by Venkat6871** [comment]
Hi @KOLANICH ,

We are closing this issue due to inactivity. If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it.

Thank you for your understanding.
